File: solutions/go/minesweeper/1/minesweeper.go
package minesweeper

import "strconv"

// Annotate returns an annotated board
func Annotate(board []string) []string {
    rowCount := len(board)

    for i := 0; i < rowCount; i++ {
        row := []rune(board[i])
        columnCount := len(row)

        for j := 0; j < columnCount; j++ {
            if !IsBomb(row[j]) { continue }

            if i > 0 {
                board[i - 1] = UpdateRow(board[i - 1], j)
            }

            board[i] = UpdateRow(board[i], j)

            if i < rowCount - 1 {
                board[i + 1] = UpdateRow(board[i + 1], j)
            }
        }
    }

    return board
}

func IsBomb(label rune) bool {
    return label == '*'
}

func UpdateRow(row string, index int) string {
    tiles := []rune(row)

    if index > 0 && !IsBomb(tiles[index - 1]) {
        tiles[index - 1] = Increment(tiles[index - 1])
    }

    if !IsBomb(tiles[index]) {
        tiles[index] = Increment(tiles[index])
    }

    if index < len(tiles) - 1 && !IsBomb(tiles[index + 1]) {
        tiles[index + 1] = Increment(tiles[index + 1])
    }

    return string(tiles)
}

func Increment(label rune) rune {
    if label >= '1' && label <= '7' {
        return []rune(strconv.Itoa(int(label - '0') + 1))[0]
    }

    return '1'
}

File: solutions/javascript/minesweeper/2/minesweeper.js
const isBomb = (c) => c === '*';

const updateRow = (row, index) => {
  let tiles = row.split('');

  if (index > 0 && !isBomb(tiles[index - 1])) {
    tiles[index - 1] = +tiles[index - 1] + 1;
  }

  if (!isBomb(tiles[index])) {
    tiles[index] = +tiles[index] + 1;
  }

  if (index < tiles.length - 1 && !isBomb(tiles[index + 1])) {
    tiles[index + 1] = +tiles[index + 1] + 1;
  }

  return tiles.join('');
};

export const annotate = (grid) => {
  for (let i = 0; i < grid.length; i++) {
    for (let j = 0; j < grid[i].length; j++) {
      if (!isBomb(grid[i].charAt(j))) continue;

      if (i > 0) {
        grid[i - 1] = updateRow(grid[i - 1], j);
      }

      grid[i] = updateRow(grid[i], j);

      if (i < grid.length - 1) {
        grid[i + 1] = updateRow(grid[i + 1], j);
      }
    }
  }

  return grid;
};
